### PR TITLE
[Ready] Peer service

### PR DESCRIPTION
**Pull request by Warchant**
## What is this pull request?
This is simple implementation of asynchronous peer service.
   
## Why do you implement it? Why do we need this pull request?
It is able to answer the following questions:
- what peers are currently online
- what is their state (state is `<merkle root, height>`)
- what is the latest state among all peers

It DOES NOT answer the questions:
- who is the leader
- what is the current order
- who is the proxy tail
These are consensus-specific things and will be implemented in consensus module. From peer service we need only availability and peer's state.
  
## How to use the features provided in the pull request?
There is an example in:
`iroha/test/module/irohad/peerservice/peerservice_peer`

1. go to build directory
2. `cmake ..`
3. `make peerservice_peer`
4. `cd test_bin`
5. ```
    peerservice_peer 4 0
    peerservice_peer 4 1
    peerservice_peer 4 2
    peerservice_peer 4 3
    ```

*Peer service is an implementation of GRPC server, it MUST be registered in grpc::ServerBuilder*.

## Additional dependencies (optional)
`lookup3.c`

# How  to make it better?
- add logger
- handle exception cases 
- validate heartbeats (hash lengths, pubkey lengths, etc) 
- add unit tests

### CODE REVIEW

**Comment by takemiyamakoto** [comment]
Yes, because modern compilers are nice 👍

File: irohad/consensus/consensus_service.hpp [hunk 1]
```diff
@@ -20,13 +20,15 @@
 
 #include <consensus.grpc.pb.h>
 #include <consensus/messages.hpp>
+#include <uvw/emitter.hpp>
 
-namespace iroha {
+namespace consensus {
 
   using ServerContext = grpc::ServerContext;
   using Status = grpc::Status;
```
Comment (motxx) [inline irohad/consensus/consensus_service.hpp]: Is it ok that `using` in header?

File: irohad/peer_service/connection_to.cpp [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+#include "connection_to.hpp"
+#include "service.hpp"
+
+namespace peerservice {
+  std::default_random_engine ConnectionTo::generator;
+  // uniform distribution for timer : from 1 sec to 2 secs
+  std::uniform_int_distribution<uint16_t> ConnectionTo::next_timer(1000, 2000);
+
+  ConnectionTo::ConnectionTo(const Node &n, std::shared_ptr<uvw::Loop> loop)
```
Comment (motxx) [inline irohad/peer_service/connection_to.cpp]: Can it be added const? `std::shared_ptr<uvw::Loop> const`

File: irohad/peer_service/connection_to.cpp [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    if (status.ok()) {
+      // peer is alive
+      this->start_timer();
+      this->online = true;
+      return answer;
+    } else {
+      if (status.error_code() == grpc::StatusCode::CANCELLED) {
+        // our heartbeat is invalid
+        // TODO handle this
+        throw "something";
```
Comment (motxx) [inline irohad/peer_service/connection_to.cpp]: This shows: `terminate called after throwing an instance of 'char const*'`
(But this is TODO, so maybe this comment can be ignored)

File: irohad/peer_service/connection_to.hpp [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+#include <grpc++/grpc++.h>
+#include <peer_service.grpc.pb.h>
+#include <memory>
+#include <uvw.hpp>
+#include "node.hpp"
+
+namespace peerservice {
+
+  struct ConnectionTo : public::uvw::Emitter<ConnectionTo> {
+    explicit ConnectionTo(const Node& n, std::shared_ptr<uvw::Loop> loop);
```
Comment (motxx) [inline irohad/peer_service/connection_to.hpp]: Maybe able to add `const` to `std::shared_ptr<>`

File: irohad/peer_service/service.cpp [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+ */
+
+#include "service.hpp"
+#include <future>
+
+namespace peerservice {
+
+  PeerServiceImpl::PeerServiceImpl(const std::vector<Node> &cluster,
+                                   const pubkey_t self,
+                                   std::shared_ptr<uvw::Loop> loop)
```
Comment (motxx) [inline irohad/peer_service/service.cpp]: Maybe able to add `const&` to std::shared_ptr<>

File: irohad/peer_service/connection_to.hpp [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+
+#include <grpc++/grpc++.h>
+#include <peer_service.grpc.pb.h>
+#include <memory>
+#include <uvw.hpp>
+#include "node.hpp"
+
+namespace peerservice {
+
+  struct ConnectionTo : public::uvw::Emitter<ConnectionTo> {
```
Comment (motxx) [inline irohad/peer_service/connection_to.hpp]: Typo? maybe
`public uvw::Emitter<ConnectionTo>`

File: irohad/peer_service/service.hpp [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+   *
+   * In future, every node may have certain topology -- "neighbours" -- a subset
+   * of full network.
+   */
+
+  class PeerServiceImpl : public uvw::Emitter<PeerServiceImpl>,
+                          public PeerService::Service {
+   public:
+    /**
+     * Event that is emitted when we find a peer with higher ledger that ours
```
Comment (motxx) [inline irohad/peer_service/service.hpp]: Is this for the system of peer in domain?

File: irohad/peer_service/service.hpp [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+#include <unordered_map>
+#include <uvw.hpp>
+#include "connection_to.hpp"
+#include "node.hpp"
+
+namespace peerservice {
+
+  /**
+   * The task of peer service is to send heartbeat messages with given interval
+   * of time to every node in the ledger.
```
Comment (motxx) [inline irohad/peer_service/service.hpp]: In this implement, where should be @satellitex 's peer service?

**Comment by Warchant** [comment]
This branch is still in *WIP*, because there is a bug (in `peerservice_peer`) with public key parsing. 

How to test it:
1. `make peerservice_peer`
2. ```
    // in different terminals
    test_bin/peerservice_peer 4 0
    test_bin/peerservice_peer 4 1
    test_bin/peerservice_peer 4 2
    test_bin/peerservice_peer 4 3
    ```

**Comment by Warchant** [comment]
TODO: there is a bug with long timer. 

Consider a system of 4 peers:
1. When we run 1 peer, wait until all timers show that peers are dead. Long timer will be started.
2. Then, one by one, run other peers.
3. First peer does not print anything (blocked at some point.) -- it is a bug and shoulf be fixed.

Working on this.

**Comment by Warchant** [comment]
It is ready for review.

File: irohad/consensus/consensus_service.hpp (hunk 1 above)
Comment (Warchant) [inline irohad/consensus/consensus_service.hpp]: Probably this is not okay. I will change it in another branch. I pushed it in this branch accidentally. Sorry :)

File: irohad/peer_service/connection_to.cpp (hunk 2 above)
Comment (Warchant) [inline irohad/peer_service/connection_to.cpp]: It can not. Loop is gonna be changed inside.

File: irohad/peer_service/connection_to.cpp [hunk 9]
```diff
@@ -59,32 +81,38 @@ namespace peerservice {
     Heartbeat answer;
 
     status = stub_->RequestHeartbeat(&context, *request, &answer);
-    printf("ping %s... ", context.peer().c_str());
+    printf("[my ledger is %d] ping %s... ", request->height(),
```
Comment (Warchant) [inline irohad/peer_service/connection_to.cpp]: All printf calls should be replaced to logger. Do we have working logger now?

File: irohad/peer_service/service.cpp (hunk 5 above)
Comment (Warchant) [inline irohad/peer_service/service.cpp]: loop will be changed inside, so no. It can not be const

File: irohad/peer_service/service.cpp [hunk 10]
```diff
@@ -37,45 +36,52 @@ namespace peerservice {
...
-    if (myHeartbeat.gmroot().length() != iroha::hash256_t::size()) {
+  void PeerServiceImpl::start() {
+    if (latestState.gmroot().length() != iroha::hash256_t::size()) {
       throw std::invalid_argument("add your heartbeat");
     }
 
     for (auto &&entry : cluster_) {
       auto &&node = entry.second;
-      node->start_timer(node->next_short_timer);
+      std::async(std::launch::async,
```
Comment (Warchant) [inline irohad/peer_service/service.cpp]: `std::async` uses system-wide thread pool for async execution.

File: test/module/irohad/peerservice/peerservice_peer.cpp [hunk 11]
```diff
@@ -64,13 +66,20 @@ int main(int argc, char** argv) {
   auto loop = uvw::Loop::create();
   PeerServiceImpl ps(cluster, cluster[ME].pubkey, hb, loop);
 
+  ps.on<Heartbeat>([self = cluster[ME].pubkey](const Heartbeat &event,
+                                               PeerServiceImpl &service) {
+    printf("\n[me: %s] Now I know that peer %s has ledger with height %d\n",
```
Comment (Warchant) [inline test/module/irohad/peerservice/peerservice_peer.cpp]: we need logger :\

**Comment by motxx** [comment]
Is this PeerService Raft? Is it used by sumeragi for electing leader?

**Comment by Warchant** [comment]
It is not raft. Sumeragi relies on order and availability of peers. This peer service does not provide the ordering, because it is consensus-dependent part (order). But availability + some extra info (useful for synchronization).

The reason why I can't start implementing sumeragi without peer service is the following:
1. consider you run N nodes. at the very beginning you don't have any leader or common order.
2. each of them has its current state (ametsuchi)
3. some nodes may have one state, other may have another. It means, that there is no common order.
4. peer A thinks that peer B is the leader, peer B thinks that peer C is the leader, and peer C is dead. 

It results in "deadlock", because nobody is sending proposals. 

To prevent this every node pings every other node and fetches its state (ledger height). And, eventually we will elect the leader who is going to share his "view" on a cluster.

**Comment by muratovv** [comment]
Why bChain ordering not applicable here?

**Pull request merged**